### PR TITLE
Interface with systemd to launch containers

### PR DESCRIPTION
**Pull request by mheon**
Adds very basic interface with systemd to allow launching of user-specified OCI bundles with runc as systemd transient services. Does not actually create the OCI bundle yet, just allows it to be passed in via CLI arg.

### CODE REVIEW

File: run_container.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package main
```
Comment (rhatdan) [inline run_container.go]: Shouldn't this be run_systemd_container?  I would think users are going to want to be able to just run regular containers with kpod also.

File: run_container.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/coreos/go-systemd/dbus"
+)
+
+// Public Functions
+
+// RunContainer runs a container with given ID
+func RunContainer(containerID string, containerPath string) error {
```
Comment (rhatdan) [inline run_container.go]: RunSystemdContainer or RunSystemContainer

File: run_container.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	return runContainerInternal(containerID, containerPath, conn)
+}
+
+// Internal Functions
+
+// Produce a per-container unique name used to identify its transient systemd service
+func getUnitName(containerID string) string {
+	return fmt.Sprintf("kpod-container-%s.service", containerID)
```
Comment (rhatdan) [inline run_container.go]: kpod should create pods not containers. At least this is what I thought we agreed on.

Comment (rhatdan) [reply run_container.go]: kpod-%s.service

File: run_container.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+// Schedule a container as a transient systemd unit
+func runContainerInternal(containerID, containerPath string, conn *dbus.Conn) error {
```
Comment (rhatdan) [inline run_container.go]: Should we build a object with conn as part of the object rather then passing this in.

File: run_container.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		return errors.New("No connection to DBus available!")
+	}
+
+	serviceName := getUnitName(containerID)
+
+	// Check if a unit with this name already exists
+	running, err := isContainerRunning(serviceName, conn)
+	if err != nil {
+		return err
+	} else if running {
```
Comment (rhatdan) [inline run_container.go]: No need for else.

File: run_container.go (hunk 3 above)
Comment (mheon) [reply run_container.go]: These are individual containers within a pod. I don't think we want to schedule an entire pod within a single systemd unit, we won't be able to track individual containers within the pod that way?

File: run_container.go (hunk 2 above)
Comment (mheon) [reply run_container.go]: These aren't system containers - we're just launching normal containers as systemd services, so we don't have to manage their lifecycle.

File: run_container.go (hunk 1 above)
Comment (grantseltzer) [reply run_container.go]: I agree here. It is idiomatic for go packages to [not](https://blog.golang.org/package-name) be snake or camel case, `systemdcontainer` would work well.

File: run_container.go (hunk 3 above)
Comment (rhatdan) [reply run_container.go]: My understanding was kpod was about using images and pods, so I definitely think you want to be able to schedule a pod within a systemd unit file. You might be running an httpd container with a monitoring agent in the same pod.  Why not schedule them together.  

The goals of these tools is to "debug" what kubernetes is doing, so I believe we should work with pods, even if 99.99% of the time they are just single containers running within the pod.

File: run_container.go (hunk 2 above)
Comment (rhatdan) [reply run_container.go]: Then call then Systemd containers.  The point I am after it we need to support the equivalent of
docker run -ti fedora sh
Which would probably execute RunContainer()  :^)

File: run_container.go (hunk 3 above)
Comment (mheon) [reply run_container.go]: I'll investigate running multiple processes in a single unit, but I don't think this is how systemd intended things to be done. My preference here would be to have a single service for each container, and then a service for the pod that depends on all the containers within it

File: run_container.go (hunk 2 above)
Comment (rhatdan) [reply run_container.go]: BTW My definition os a system container is just a container launched from a systemd unit file using runc.  But maybe other people here that name and thing of something different.

And to go with our other discussion it should probably be RunSystemdPod.

If we move to the Pod Method then we will need a way to join a Container to an existing POD.

File: run_container.go (hunk 4 above)
Comment (mheon) [reply run_container.go]: Maybe, but I don't think there's a need yet. I don't anticipate having much code here, most of the heavy lifting will be done with runc

File: run_container.go (hunk 2 above)
Comment (mheon) [reply run_container.go]: I knew system containers were systemd-related so I was assuming we were using nspawn - oops. These are definitely system containers, then.

**Comment by mrunalp** [comment]
I think we are approaching this the wrong direction. This could probably be added later once we build up the basic features and CLI first.

File: run_container.go (hunk 2 above)
Comment (mrunalp) [reply run_container.go]: @mheon Take a look at ocid to see how pods are created. We will be taking a similar approach and use ns fds for representing a pod here.

**Comment by mheon** [comment]
I was going at this bottom-up - start with running individual containers, then creating and running them, then creating and running whole pods, then build a CLI on top of that. If you'd prefer I go about it top-down, I'd be happy to switch approach.

Regardless, I'll push a new version in a few minutes with fixes. Renaming and moving run_container into a package means I actually have to worry about gopath again, so it broke some things.

File: run_container.go (hunk 2 above)
Comment (mheon) [reply run_container.go]: @mrunalp That's not dissimilar to what I was originally thinking - but I'm not viewing this as part of the pod interface. There's an interface for talking directly to runc in ocid to get an individual container within a pod spooled up - that's what I was trying to produce here. The pod abstraction can be built once we have the ability to get containers started/stopped (unless you're saying we should skip containers and just try and handle whole pods at a time?)

**Comment by mheon** [comment]
Renamed run_container to system_container (along with associated functions), and moved it into a package

**Pull request closed without merging**